File: worker/src/athlete.js
// athlete.js — the athlete/player profile tier (canonical AthleteProfile;
// SCORES-APP-BUILD-SPEC §2.6 "Player rows"). Pure map→map, no I/O, so it runs in
// Node tests + the Dart port alike. This is the RICH, lazy, on-open detail for one
// player — identity + season stats + a last-N game log — NEVER on the cheap
// scoreboard poll. The caller (api.dart) does the CORE fetches + the $ref fan-out
// and hands this normalizer the already-resolved raws; this file only shapes them.
//
// Inputs (all pre-resolved by the caller):
//   • identity   — the roster row (denser, single-call when arriving from a team)
//                  OR the core athletes/{id} doc. Both share the same field shape.
//   • team       — the resolved team.$ref doc (name/color/logo), or null.
//   • statistics — the resolved athletes/{id}/statistics doc, or null.
//   • games      — [{ eventId, teamId, event, statistics }] — the last-N eventlog
//                  items with their event.$ref + statistics.$ref pre-resolved
//                  (either may be null on a failed resolve). Most-recent first.
//
// EVIDENCE: every path was OBSERVED in schema/espn-guide/core-athletes-id.md /
// -statistics.md / -eventlog.md and verified against a live probe (MLB + WNBA,
// 2026-07). Per-game stat NAMES are inferred in the guide → we bind to whatever the
// split carries, never fabricate. Fields not observed (see §2.6 Gaps) are omitted.

const https = (u) => (typeof u === 'string' ? u.replace(/^http:/, 'https:') : undefined);

// Dark-mode logo: explicit ESPN 'dark' rel when present, else the /500/→/500-dark/
// CDN derivation. Same rule as standings.js / rankings.js (no sport gate).
function darkLogoOf(team) {
  const ls = team?.logos;
  if (Array.isArray(ls)) {
    const d = ls.find((l) => (l.rel || []).includes('dark') && !(l.rel || []).includes('scoreboard'))
      || ls.find((l) => (l.rel || []).includes('dark'));
    if (d?.href) return https(d.href);
  }
  const light = https(ls?.[0]?.href);
  return (light && light.includes('/i/teamlogos/') && light.includes('/500/'))
    ? light.replace('/500/', '/500-dark/')
    : undefined;
}

// headshot.href (or a bare string) → https. Both the roster row and the core doc
// carry `headshot: { href, alt }`.
function headshotOf(o) {
  return https(o?.headshot?.href || o?.headshot);
}

// position.abbreviation (preferred — 'RP'/'G') falling back to displayName. Both
// identity sources carry the same position object.
function positionOf(o) {
  const p = o?.position;
  if (!p || typeof p !== 'object') return undefined;
  return p.abbreviation || p.displayName || undefined;
}

// splits.categories[] → compact [{ name, displayName, stats: [cell] }]. Shared by
// season totals and the per-game line (identical ESPN shape). Drops the verbose
// `description`; keeps only cells with a `name` and a `displayValue`. Categories
// with no usable cells are dropped; returns undefined when nothing survives.
function buildStatCategories(statsDoc) {
  const cats = statsDoc?.splits?.categories;
  if (!Array.isArray(cats) || !cats.length) return undefined;
  const out = [];
  for (const c of cats) {
    const cells = [];
    for (const s of Array.isArray(c?.stats) ? c.stats : []) {
      if (!s || !s.name || s.displayValue == null) continue;
      const cell = { name: String(s.name), displayValue: String(s.displayValue) };
      if (s.abbreviation != null && s.abbreviation !== '') cell.abbreviation = String(s.abbreviation);
      if (s.displayName != null && s.displayName !== '') cell.displayName = String(s.displayName);
      if (s.shortDisplayName != null && s.shortDisplayName !== '') cell.shortDisplayName = String(s.shortDisplayName);
      if (typeof s.value === 'number' && Number.isFinite(s.value)) cell.value = s.value;
      cells.push(cell);
    }
    if (!cells.length) continue;
    const cat = { name: String(c.name || c.abbreviation || ''), stats: cells };
    if (c.displayName != null && c.displayName !== '') cat.displayName = String(c.displayName);
    out.push(cat);
  }
  return out.length ? out : undefined;
}

// One last-N row from a resolved eventlog item. eventId/teamId are already strings
// on the item; date/name/shortName come from the resolved event; the per-game line
// from the resolved statistics. A wholly-unresolved row (no event, no stats) still
// keeps its id so the app can link it.
function buildGameRow(g) {
  if (!g || g.eventId == null) return null;
  const row = { eventId: String(g.eventId) };
  const ev = g.event;
  if (ev && typeof ev === 'object') {
    if (typeof ev.date === 'string' && ev.date) row.date = ev.date;
    if (typeof ev.name === 'string' && ev.name) row.name = ev.name;
    if (typeof ev.shortName === 'string' && ev.shortName) row.shortName = ev.shortName;
  }
  if (g.teamId != null && g.teamId !== '') row.teamId = String(g.teamId);
  const stats = buildStatCategories(g.statistics);
  if (stats) row.stats = stats;
  return row;
}

// team.$ref doc → the athlete's team block. null-safe: no doc → undefined.
function buildTeam(team) {
  if (!team || typeof team !== 'object' || team.id == null) return undefined;
  const out = {
    id: String(team.id),
    name: team.displayName || team.name || team.shortDisplayName || '',
  };
  if (team.abbreviation) out.abbr = String(team.abbreviation);
  if (team.color) out.color = String(team.color);
  const logo = https(team?.logos?.[0]?.href);
  if (logo) out.logo = logo;
  const dark = darkLogoOf(team);
  if (dark) out.logoDark = dark;
  return out;
}

/**
 * Compose a canonical AthleteProfile from the pre-resolved CORE inputs.
 * @param {string} league   the league key (e.g. 'baseball/mlb')
 * @param {string} athleteId
 * @param {object} parts    { identity, team, statistics, games }
 */
export function normalizeAthleteProfile(league, athleteId, parts = {}) {
  const { identity, team, statistics, games } = parts;
  const idn = identity && typeof identity === 'object' ? identity : {};
  const out = {
    id: String(idn.id != null ? idn.id : athleteId),
    league: String(league),
    name: idn.displayName || idn.fullName || idn.shortName || '',
  };
  if (idn.shortName) out.shortName = String(idn.shortName);
  if (idn.jersey != null && idn.jersey !== '') out.jersey = String(idn.jersey);
  const pos = positionOf(idn);
  if (pos) out.position = pos;
  const hs = headshotOf(idn);
  if (hs) out.headshot = hs;
  if (typeof idn.age === 'number' && Number.isFinite(idn.age)) out.age = idn.age;
  if (typeof idn.displayHeight === 'string' && idn.displayHeight) out.height = idn.displayHeight;
  if (typeof idn.displayWeight === 'string' && idn.displayWeight) out.weight = idn.displayWeight;

  const tm = buildTeam(team);
  if (tm) out.team = tm;

  const stats = buildStatCategories(statistics);
  if (stats) out.stats = stats;

  if (Array.isArray(games) && games.length) {
    const rows = games.map(buildGameRow).filter(Boolean);
    if (rows.length) out.lastGames = rows;
  }
  return out;
}

File: app/lib/src/data/athlete.dart
// athlete.dart — Dart port of worker/src/athlete.js. The athlete/player profile
// tier (canonical AthleteProfile; SCORES-APP-BUILD-SPEC §2.6). Pure map→map,
// verified byte-for-byte against the JS oracle via test/port_athlete_test.dart.
//
// The RICH, lazy, on-open detail for one player — identity + season stats + a
// last-N game log. NEVER on the cheap scoreboard poll: the caller (api.dart) does
// the CORE fetches + the $ref fan-out and hands this normalizer the pre-resolved
// raws; this file only shapes them. Every path is OBSERVED in the espn-guide
// core-athletes pages + a live probe (MLB/WNBA, 2026-07); not-observed fields are
// omitted, never faked. See the JS oracle header for the full input contract.

import 'util.dart';

// Dark-mode logo: explicit 'dark' rel, else the /500/→/500-dark/ derivation. Same
// rule as standings/rankings — util.dart's darkFromLogos is the shared port.
String? _darkLogoOf(dynamic team) => darkFromLogos(team);

// headshot.href (or bare string) → https.
String? _headshotOf(dynamic o) => https(field(field(o, 'headshot'), 'href') ?? field(o, 'headshot'));

// position.abbreviation (preferred) falling back to displayName.
String? _positionOf(dynamic o) {
  final p = field(o, 'position');
  if (p is! Map) return null;
  final abbr = p['abbreviation'];
  if (abbr != null && abbr != '') return abbr.toString();
  final dn = p['displayName'];
  if (dn != null && dn != '') return dn.toString();
  return null;
}

// splits.categories[] → compact [{name, displayName?, stats:[cell]}]. Shared by
// season totals and the per-game line. Drops `description`; keeps only cells with a
// name + displayValue. Empty categories dropped; null when nothing survives.
List<Map<String, dynamic>>? _buildStatCategories(dynamic statsDoc) {
  final cats = field(field(statsDoc, 'splits'), 'categories');
  if (cats is! List || cats.isEmpty) return null;
  final out = <Map<String, dynamic>>[];
  for (final c in cats) {
    final rawStats = field(c, 'stats');
    final cells = <Map<String, dynamic>>[];
    for (final s in rawStats is List ? rawStats : const []) {
      if (s is! Map) continue;
      final name = s['name'];
      if (name == null || name == '' || s['displayValue'] == null) continue;
      final cell = <String, dynamic>{
        'name': name.toString(),
        'displayValue': s['displayValue'].toString(),
      };
      if (s['abbreviation'] != null && s['abbreviation'] != '') cell['abbreviation'] = s['abbreviation'].toString();
      if (s['displayName'] != null && s['displayName'] != '') cell['displayName'] = s['displayName'].toString();
      if (s['shortDisplayName'] != null && s['shortDisplayName'] != '') cell['shortDisplayName'] = s['shortDisplayName'].toString();
      final v = s['value'];
      if (v is num && v.isFinite) cell['value'] = v;
      cells.add(cell);
    }
    if (cells.isEmpty) continue;
    final cat = <String, dynamic>{
      'name': or([field(c, 'name'), field(c, 'abbreviation'), '']).toString(),
      'stats': cells,
    };
    final dn = field(c, 'displayName');
    if (dn != null && dn != '') cat['displayName'] = dn.toString();
    out.add(cat);
  }
  return out.isEmpty ? null : out;
}

// One last-N row from a resolved eventlog item {eventId, teamId, event, statistics}.
Map<String, dynamic>? _buildGameRow(dynamic g) {
  if (g is! Map || g['eventId'] == null) return null;
  final row = <String, dynamic>{'eventId': g['eventId'].toString()};
  final ev = g['event'];
  if (ev is Map) {
    if (ev['date'] is String && (ev['date'] as String).isNotEmpty) row['date'] = ev['date'];
    if (ev['name'] is String && (ev['name'] as String).isNotEmpty) row['name'] = ev['name'];
    if (ev['shortName'] is String && (ev['shortName'] as String).isNotEmpty) row['shortName'] = ev['shortName'];
  }
  if (g['teamId'] != null && g['teamId'] != '') row['teamId'] = g['teamId'].toString();
  final stats = _buildStatCategories(g['statistics']);
  if (stats != null) row['stats'] = stats;
  return row;
}

// team.$ref doc → the athlete's team block. null-safe.
Map<String, dynamic>? _buildTeam(dynamic team) {
  if (team is! Map || team['id'] == null) return null;
  final out = <String, dynamic>{
    'id': team['id'].toString(),
    'name': or([team['displayName'], team['name'], team['shortDisplayName'], '']),
  };
  if (truthy(team['abbreviation'])) out['abbr'] = team['abbreviation'].toString();
  if (truthy(team['color'])) out['color'] = team['color'].toString();
  final logo = https(field(first(field(team, 'logos')), 'href'));
  if (logo != null) out['logo'] = logo;
  final dark = _darkLogoOf(team);
  if (dark != null) out['logoDark'] = dark;
  return out;
}

/// Compose a canonical AthleteProfile from the pre-resolved CORE inputs.
/// [parts] = {identity, team, statistics, games}.
Map<String, dynamic> normalizeAthleteProfile(
    String league, String athleteId, Map<String, dynamic> parts) {
  final identity = parts['identity'];
  final idn = identity is Map ? identity : const {};
  final out = <String, dynamic>{
    'id': (idn['id'] ?? athleteId).toString(),
    'league': league,
    'name': or([idn['displayName'], idn['fullName'], idn['shortName'], '']),
  };
  if (truthy(idn['shortName'])) out['shortName'] = idn['shortName'].toString();
  if (idn['jersey'] != null && idn['jersey'] != '') out['jersey'] = idn['jersey'].toString();
  final pos = _positionOf(idn);
  if (pos != null) out['position'] = pos;
  final hs = _headshotOf(idn);
  if (hs != null) out['headshot'] = hs;
  final age = idn['age'];
  if (age is num && age.isFinite) out['age'] = age;
  if (idn['displayHeight'] is String && (idn['displayHeight'] as String).isNotEmpty) {
    out['height'] = idn['displayHeight'];
  }
  if (idn['displayWeight'] is String && (idn['displayWeight'] as String).isNotEmpty) {
    out['weight'] = idn['displayWeight'];
  }

  final tm = _buildTeam(parts['team']);
  if (tm != null) out['team'] = tm;

  final stats = _buildStatCategories(parts['statistics']);
  if (stats != null) out['stats'] = stats;

  final games = parts['games'];
  if (games is List && games.isNotEmpty) {
    final rows = <Map<String, dynamic>>[];
    for (final g in games) {
      final r = _buildGameRow(g);
      if (r != null) rows.add(r);
    }
    if (rows.isNotEmpty) out['lastGames'] = rows;
  }
  return out;
}
